### PR TITLE
Ignore default DB_FILENAME option when using MySQL/MariaDB

### LINKED ISSUE #22969: v11.0.0-rc3. In Docker, 'filename' env is incompatible with MySQL2
### Describe the Bug

This is the warning in the console when using mariadb:
```terminal
Ignoring invalid configuration option passed to Connection: filename. This is currently a warning, but in future versions of MySQL2, an error will be thrown if you pass an invalid configuration option to a Connection
```

### Directus Version

`v11.0.0-rc.3`

### Hosting Strategy

Docker

### Database

MariaDB

### PR DESCRIPTION
**Pull request by cliqer**
## Scope

What's changed:

- Conditionally set DB_FILENAME if DB_CLIENT is sqlite.
This fixes the MySQL2 Warning in docker:
```
Ignoring invalid configuration option passed to Connection: filename. This is currently a warning, but in future versions of MySQL2, an error will be thrown if you pass an invalid configuration option to a Connection
```

## Potential Risks / Drawbacks

- none

## Review Notes / Questions

---

Fixes #22969

### CODE REVIEW

**Review by paescuj — CHANGES_REQUESTED** [review]
That won't work, as the `RUN` command is part of the build process and not applied on runtime.

Let's drop the config on the following line instead: https://github.com/directus/directus/blob/178ee402a6e2df050f7ed81e77983a28a12e54cb/api/src/database/index.ts#L139

**Comment by cliqer** [comment]
This also needs to be dropped in the Dockerfile as it is always added even when not needed.
i.e. running with plesk docker there is no way to remove the warning if it is not conditionally loaded in the dockerfile.

**Comment by rijkvanzanten** [comment]
@cliqer Instead of removing it from the Dockerfile, we can drop any unknown configuration flags at the time where we use the environment variables. That can be done here:

https://github.com/directus/directus/blob/178ee402a6e2df050f7ed81e77983a28a12e54cb/api/src/database/index.ts#L139

🙂

**Comment by br41nslug** [comment]
> Instead of removing it from the Dockerfile, we can drop any unknown configuration flags at the time where we use the environment variables.

There is quite a list of options to check against but feasible to do this using an allow list of `ConnectionOptions` for mysql https://github.com/sidorares/node-mysql2/blob/master/typings/mysql/lib/Connection.d.ts#L82

**Comment by rijkvanzanten** [comment]
There's an alternative @br41nslug: Instead of adding those env vars in Dockerfile, we could have Directus default to SQLite with that path if no DB env vars are passed at all. That avoids having to have an allow list for MySQL specifically and simplifies our dockerfile a bit as well

**Comment by br41nslug** [comment]
Yeah would much prefer using the defaults in `@directus/env` over setting default env values in the dockerfile 😄

**Comment by paescuj** [comment]
Why not simply dropping `filename` in 
https://github.com/directus/directus/blob/178ee402a6e2df050f7ed81e77983a28a12e54cb/api/src/database/index.ts#L139? This seems to be the only conflicting option, while I think we want to set this as the default specifically for Docker context where we also have the `database` folder prepared.

**Comment by rijkvanzanten** [comment]
@paescuj Why not simply use defaults for those env vars so it's consistent with or without Docker? 😁 

Adding those manual exceptions feels a little brittle to me, and so does trying to maintain an allow list of mysql options 🤔

**Comment by paescuj** [comment]
@rijkvanzanten We can sure do, it just doesn't seem like the straight-forward solution here.
Defaults via `@directus/env` would required additional logic in this package, as we would need to _conditionally_ apply the `DB_FILENAME` default, depending on the value of `DB_CLIENT`, otherwise we run into the same problem.
Also the default of `database/database.sqlite` makes sense for the Docker image, but does it in non-Docker context where the `database` folder needs to be created manually?

The warning is only about `filename`. If there should be any other conflicting options, these would have been explicitly defined by the admin, which means the admin can easily remove/correct those. Therefore, I don't think we need to have put an allow list in place.

IMO, it could be a simple as: 

```diff
+		if (isObject(knexConfig.connection)) delete knexConfig.connection['filename'];
+
		Object.assign(knexConfig, { client: 'mysql2' });
```

We're already doing some other config adjustments in this file like https://github.com/directus/directus/blob/4fcc7304062afe2909d359fac0715f4e7b8bf258/api/src/database/index.ts#L154 so I feel like this additional line is acceptable.

**Comment by rijkvanzanten** [comment]
Fair enough! Lets do that 👍🏻

File: api/src/database/index.ts [hunk 1]
```diff
@@ -136,8 +136,10 @@ export function getDatabase(): Knex {
 	}
 
 	if (client === 'mysql') {
+		if (isObject(knexConfig.connection)) delete knexConfig.connection['filename'];
+		
 		Object.assign(knexConfig, { client: 'mysql2' });
-
+
```
Comment (hanneskuettner) [inline api/src/database/index.ts:142]: Looks like some tabs got in here by accident

File: Dockerfile [hunk 2]
```diff
@@ -53,7 +53,7 @@ EXPOSE 8055
 
 ENV \
 	DB_CLIENT="sqlite3" \
-	DB_FILENAME="/directus/database/database.sqlite" \
+ 	DB_FILENAME="/directus/database/database.sqlite" \
```
Comment (paescuj) [inline Dockerfile:56]:
```suggestion
	DB_FILENAME="/directus/database/database.sqlite" \
```

File: api/src/database/index.ts [hunk 3]
```diff
@@ -136,8 +136,9 @@ export function getDatabase(): Knex {
 	}
 
 	if (client === 'mysql') {
+		if (isObject(knexConfig.connection)) delete knexConfig.connection['filename'];
```
Comment (paescuj) [inline api/src/database/index.ts:139]: `isObject` needs to be imported from `@directus/utils`

Comment (paescuj) [reply api/src/database/index.ts:139]: Let's also add a small comment:
```suggestion
		// Remove the conflicting `filename` option, defined by default in the Docker Image
		if (isObject(knexConfig.connection)) delete knexConfig.connection['filename'];
```

File: api/src/database/index.ts [hunk 4]
```diff
@@ -136,8 +136,9 @@ export function getDatabase(): Knex {
...
 		Object.assign(knexConfig, { client: 'mysql2' });
-
+
```
Comment (paescuj) [inline api/src/database/index.ts:141]:
```suggestion

```

File: .changeset/ten-beds-pretend.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"@directus/api": patch
+---
+
+Ensured the default `DB_FILENAME` option from the Docker Image is not applied when using MySQL/MariaDB, which fixes a corresponding warning
```
Comment (paescuj) [inline .changeset/ten-beds-pretend.md:5]:
```suggestion
Ensured the default `DB_FILENAME` option from the Docker Image is not applied when using MySQL/MariaDB, fixing a corresponding warning
```

**Review by paescuj — APPROVED** [review]
Thanks @cliqer!

**Pull request merged**